### ISSUE TITLE
dry-initializer: `to_s': wrong number of arguments (1 for 0)

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1462]
This failed under `truffleruby-1.0.0.rc9`. No issue under MRI.

```ruby
require 'dry-initializer'

class Test
  extend Dry::Initializer

  param :foo, proc(&:to_s)
end

Test.new(:FOO)
```

Error:

```
dry-initializer.rb:6:in `to_s': wrong number of arguments (1 for 0) (ArgumentError)
	from dry-initializer.rb:6:in `to_s'
	from (eval):5:in `call'
	from (eval):5:in `__dry_initializer_initialize__'
	from ~/.gem/truffleruby/2.4.4/gems/dry-initializer-2.5.0/lib/dry/initializer/mixin/root.rb:7:in `initialize'
	from dry-initializer.rb:9:in `new'
	from dry-initializer.rb:9:in `<main>'
```

**Comment by chrisseaton** [comment]
Thanks - we'll have to dig into the `dry` stack a bit for this.

**Comment by deepj** [comment]
@eregon Is possible to look at this issue, please? This issue affects the whole dry-rb/ROM ecosystem. It's some ugly issue with metaprogramming,

**Comment by eregon** [comment]
@deepj This variant seems to work fine:
```ruby
require 'dry-initializer'

class Test
  extend Dry::Initializer

  # param :foo, proc(&:to_s)
  param :foo, &:to_s
end

p Test.new(:FOO)
```

**Comment by eregon** [comment]
Mmh, but that's different as it passes a block argument to `param`.
```ruby
require 'dry-initializer'

class Test
  extend Dry::Initializer

  # param :foo, proc(&:to_s)
  param :foo, :to_s.to_proc
end

p Test.new(:FOO)
```
also fails.

**Comment by eregon** [comment]
This works:
```ruby
require 'dry-initializer'

class Test
  extend Dry::Initializer

  param :foo, -> obj { obj.to_s }
end

p Test.new(:FOO)
p Test.new(:FOO).foo # => "FOO"
```
The code generated by dry-initializer [here](https://github.com/dry-rb/dry-initializer/blob/9e2db80dac2e08eddd27066c0ec5ad34c1c6fef6/lib/dry/initializer/builders/attribute.rb#L69-L72) differs based on the Proc#arity of a `Symbol#to_proc` Proc. TruffleRuby currently returns -2, MRI -1.

**Comment by eregon** [comment]
@deepj  Fix coming.
FWIW, an arity of -2 would better express the number of required arguments: `Symbol#to_proc` needs at least one argument (the receiver).
An arity of -1 means "0 or more arguments".
`:to_s.to_proc.call` gives `ArgumentError: no receiver given` on MRI, showing at least one argument is indeed required.

So it might be a MRI bug, on which `dry-initializer` depends. I'll report it to them.
In any case, I'll follow MRI semantics for now for compatibility.

**Comment by eregon** [comment]
@deepj Fixed, it should be in the next release. Thank you for the report and the ping!

**Comment by deepj** [comment]
@eregon thank you! it's great this has been fixed. I can't wait on running dry-rb/ROM on TruffleRuby. If you would have another time, this is another issue in that ecosystems. https://github.com/oracle/truffleruby/issues/1469

I'm not sure how this can be relative to this issue.
